File: protected_service/src/middleware/auth_middleware.ts
import {Context} from "koa";


async function authMiddleware(ctx: Context, next: Awaited<Function>) {
    const serviceExchangeKey = process.env.SERVICE_EXCHANGE_KEY || ''
    const authorizationHeaderName = process.env.SERVICE_EXCHANGE_HEADER_NAME || ''
    const authorizationHeader = ctx.request.headers[authorizationHeaderName.toLowerCase()]

    if (!authorizationHeader || authorizationHeader !== `Bearer ${serviceExchangeKey}`) {
        ctx.throw(401);
    }

    await next()
}

export default authMiddleware;

File: auth_service/core/middlewares/auth_middleware.py
from fastapi import Request
from starlette.responses import Response
from starlette.status import HTTP_401_UNAUTHORIZED

import core.settings as settings


async def auth_middleware(request: Request, call_next):
    authorization_header = request.headers.get(settings.SERVICE_EXCHANGE_HEADER_NAME, None)
    invalid_authorization_header = not authorization_header or authorization_header != f'Bearer {settings.SERVICE_EXCHANGE_KEY}'
    if request.url.path != '/api/authenticate' and invalid_authorization_header:
        return Response(status_code=HTTP_401_UNAUTHORIZED, content='UNAUTHORIZED')

    response = await call_next(request)
    return response
